[DEV APPROVED]7228 link to mas bug

misaka opened the pull request:
The tests added here include a certain amount of refactoring to partially take out the reliance on a singleton object `cms_page`. This allows us to work with two pages in the same test, for example when linking to pages like in this change. The end goal should probably be to move the fixture generation into factory girl ... although that's probably for later change.

rnalexander commented:
:+1:

tomas-stefano commented:
👍 (After fixing the code climate warning)

jiannis commented on features/step_definitions/edit_page_steps.rb:
```diff
@@ -142,3 +163,18 @@
+
+Then(/^the "([a-zA-Z]{2})" and "([a-zA-Z]{2})" versions are available for linking$/) do |*locales|
+  locales.each do |locale|
+    expect(@linkable_page.send(locale)).
```
Using `@linkable_page` as an instance variable initialised in the previous step makes these two step definitions dependent. Maybe a helper method instead of a `@linkable_page` instance variable?

misaka replied:
This is a tricky one. Not having steps dependant is good design, but in this case it gets awkward if we want to bundle both tests together into `I should see "(.*?)" in the list of linkable pages` and `the "([a-zA-Z]{2})" and "([a-zA-Z]{2})" versions are available for linking` steps. The step wording gets awkward, etc. And it's nice to test this criteria separately instead of squishing them into the same step.

jiannis replied:
👌

jiannis commented on features/support/env.rb:
```diff
@@ -8,7 +8,7 @@
 
 Capybara.register_driver :poltergeist do |app|
-  Capybara::Poltergeist::Driver.new(app, js_errors: true, inspector: true)
+  Capybara::Poltergeist::Driver.new(app, js_errors: true, inspector: true, window_size: [2048, 1536])
```
Why do need this specific window size?

misaka replied:
Only to be able to see more of the page when saving a screenshot for debugging.

jiannis replied:
👌

jiannis commented on features/support/hooks.rb:
```diff
@@ -0,0 +1,17 @@
+    end
+  end
+end
+
```
🎩

jiannis commented on features/support/world/cms.rb:
```diff
@@ -13,11 +13,15 @@ def cms_sites
 
     def cms_root(locale='en')
       self.root ||= cms_site(locale).pages.create!(layout: cms_layout, label: 'root')
+      self.root
```
Doesn't the previous statement return `self.root` ?

jiannis commented on app/controllers/links/pages_controller.rb:
```diff
@@ -1,6 +1,6 @@
 class Links::PagesController < PagesController
   def index
-    pages  = Comfy::Cms::Page.includes(:site).published.filter(params.slice(:category))
+    pages  = Comfy::Cms::Page.includes(:site).where('state != ?', 'unpublished').filter(params.slice(:category))
```
Could we do the same thing with a `scope` in [`Comfy::Cms::Page`](https://github.com/moneyadviceservice/cms/blob/master/app/models/comfy/cms/page.rb) ?

```ruby
class Comfy::Cms::Page < ActiveRecord::Base
  scope :allthethings, -> { where.not(state: 'unpublished') }
end

def index
  pages = Comfy::Cms::Page.includes(:site).allthethings.filter(params.slice(:category)
end
```

or maybe use the existing `unpublished` scope and try something like:

```ruby
def index
  pages = Comfy::Cms::Page.includes(:site).where.not.published.filter(params.slice(:category)
end
```

misaka replied:
Looks good! Let me test this.

misaka commented:
@tomas-stefano re: the [Code Climate issue in `UI::Pages::Edit`](https://codeclimate.com/repos/55798a3869568024020134d1/UI::Pages::Edit?from_sha=cad65949&to_sha=26d5331d), after reviewing the change we discussed, I think it's better to leave this as-is. As we saw, sections are stored in a separate directory in our structure, and if they're not being re-used I don't really see enough benefit to warrant moving them so far away from the code where they're being used. Do you agree?

tomas-stefano commented:
I agree in terms. The structure it is not flexible and is totally awful. Sections belongs to pages, so it should not be separated. And there is the tag section duplication either on tag.rb and header.rb. Confusing. **But** we should create a card to improve this structure (or add this info on the factory girl replacement on the cucumber spec). If this is added on the user story, I am totally fine. Otherwise, I am not.

misaka commented:
@tomas-stefano I've cleaned up the code duplication that we saw in Code Climate, and added a couple TP tickets: [7351](https://moneyadviceservice.tpondemand.com/entity/7351) and [7353](https://moneyadviceservice.tpondemand.com/entity/7353). Of course, Code Climate is still bugging about some other duplication but I reckon that that can wait until the fixture refactoring work.

tomas-stefano commented:
👍 👍

rnalexander commented:
👍  :shipit:

jongilbraith commented:
@misaka nice job with the tests, I was frustrated with the state of them last time I was working on this but didn't have it in me to do much to improve them. Still plenty of scope for improvement but your changes address the main sticking points nicely.

👍 from me.

UmairIft commented:
All looks good 👍
had a mini discussion with @jongilbraith , the issue with scheduled publishing had to do with UTC time zones. The scheduled time is -1hr of current time . A fix for another ticket
